concurrency/AnnotationStrategyHelper.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ */
+package concurrency;
+
+import com.amazon.correios.notification.component.runtime.ComponentContext;
+import com.amazon.traffic.cafe.component.annotator.strategy.AnnotationStrategy;
+import com.amazon.traffic.cafe.component.annotator.strategy.AnnotationStrategyFactory;
+import com.amazon.traffic.cafe.component.annotator.strategy.AnnotationStrategyType;
+import lombok.experimental.UtilityClass;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+
+/**
+ * Utility class to fetch AnnotationStrategy object for the given annotationStrategy type,
+ * marketplaceId. It uses concurrent HashMap to hold the instance of RODBAccess.
+ * This ensures that we don't create the multiple instance of RODBAccess for same strategy type.
+ */
+@UtilityClass
+public class AnnotationStrategyHelper {
+
+    /***
+     * map to hold the instance of AnnotationStrategy. This ensures that we don't create
+     * the multiple instance of RODBAccess for same strategy type.
+     */
+    private Map<String, AnnotationStrategy> annotationStrategyMap =
+            new ConcurrentHashMap<String, AnnotationStrategy>();
+
+    /**
+     * Annotator strategy key format to be added in the concurrent hashmap for
+     * annotation strategies. It is a combination of strategy type and
+     * marketplace.
+     */
+    private static final String ANNOTATOR_STRATEGY_KEY_FORMAT = "%s-%s";
+
+    /**
+     * Get AnnotationStrategy for the given annotation strategy type and marketplace.
+     *
+     * @param annotationStrategy
+     *              annotation strategy type
+     * @param marketplaceId
+     *              marketplace id
+     * @param context
+     *              ComponentContext object
+     * @return AnnotationStrategy object
+     */
+    public AnnotationStrategy getAnnotationStrategy(final String annotationStrategy,
+                                                    final String marketplaceId,
+                                                    final ComponentContext context) {
+
+        AnnotationStrategyType annotationStrategyType = AnnotationStrategyType.valueOf(annotationStrategy);
+        String annotatorKey = String.format(ANNOTATOR_STRATEGY_KEY_FORMAT, annotationStrategy, marketplaceId);
+        AnnotationStrategy annotator = annotationStrategyMap.get(annotatorKey);
+
+        if (annotator == null) {
+            synchronized (AnnotationStrategyHelper.class) {
+                annotator = annotationStrategyMap.get(annotatorKey);
+                if (annotator == null) {
+                    annotator = AnnotationStrategyFactory.getFilterStrategy(annotationStrategyType, marketplaceId,
+                            context);
+                    annotationStrategyMap.put(annotatorKey, annotator);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **59 and 63**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization. The check is at line: **60**.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).